receiver/awscontainerinsightreceiver/internal/host/hostinfo.go

```diff
@@ -0,0 +1,169 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package host
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/internal/aws/awsutil"
+)
+
+// Info contains information about a host
+type Info struct {
+	cancel           context.CancelFunc
+	logger           *zap.Logger
+	awsSession       *session.Session
+	refreshInterval  time.Duration
+	instanceIDReadyC chan bool // close of this channel indicates instance ID is ready
+
+	ebsVolumeReadyC chan bool // close of this channel indicates ebsVolume is initialized. It is used only in test
+	ec2TagsReadyC   chan bool // close of this channel indicates ec2Tags is initialized. It is used only in test
+
+	nodeCapacity nodeCapacityProvider
+	ec2Metadata  ec2MetadataProvider
+	ebsVolume    ebsVolumeProvider
+	ec2Tags      ec2TagsProvider
+
+	awsSessionCreator   func(*zap.Logger, awsutil.ConnAttr, *awsutil.AWSSessionSettings) (*aws.Config, *session.Session, error)
+	nodeCapacityCreator func(*zap.Logger, ...nodeCapacityOption) (nodeCapacityProvider, error)
+	ec2MetadataCreator  func(context.Context, *session.Session, time.Duration, chan bool, *zap.Logger, ...ec2MetadataOption) ec2MetadataProvider
+	ebsVolumeCreator    func(context.Context, *session.Session, string, string, time.Duration, *zap.Logger, ...ebsVolumeOption) ebsVolumeProvider
+	ec2TagsCreator      func(context.Context, *session.Session, string, time.Duration, *zap.Logger, ...ec2TagsOption) ec2TagsProvider
+}
+
+type machineInfoOption func(*Info)
+
+// NewInfo creates a new Info struct
+func NewInfo(refreshInterval time.Duration, logger *zap.Logger, options ...machineInfoOption) (*Info, error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	mInfo := &Info{
+		cancel:           cancel,
+		refreshInterval:  refreshInterval,
+		instanceIDReadyC: make(chan bool),
+		logger:           logger,
+
+		awsSessionCreator:   awsutil.GetAWSConfigSession,
+		nodeCapacityCreator: newNodeCapacity,
+		ec2MetadataCreator:  newEC2Metadata,
+		ebsVolumeCreator:    newEBSVolume,
+		ec2TagsCreator:      newEC2Tags,
```

Comment:
These feel weird to me.  Rather than having initializers for dependencies as member variables on this type I would take the dependencies directly in this initializer.  These are only used for overriding the default initializers with test doubles and the tests could provide their own value as needed.

I won't hold this PR up for this, but it would be good if you could open an issue to track replacing this with injecting the dependencies via `options` and validating that all dependencies were provided.

Reply:
I will talk to Aneurysm9 offline to address this issue. For now, let's get the PR merged as soon as possible.